### PR TITLE
Move proof generation to the type system level

### PR DESCRIPTION
**Pull request by bkchr**
This pr changes the way proof generation can be requested when building a block with `Proposer`. Instead of passing `RecordProof::Yes` to every call of `propose`, this pr changes it to either always record the proof or never for a specific implementation of `Proposer`. Up to now there is was no middle way of "now we need a proof" and "now we don't need a proof". It was always either one of this. 

The main advantage of moving this to the type system level is that we can write code `Proposer<B, ProofRecording = EnableProofRecording>`. Meaning we will only accept proposer implementations that will generate a proof. This is quite useful for Cumulus and for people who will be using Cumulus, to make it complain at compile time that the proposer is not correct.

The basic authorship `ProposerFactory` and `Proposer` are now written in a way to make them usable for proof recording or for no proof recording. 

polkadot companion: https://github.com/paritytech/polkadot/pull/2507

### CODE REVIEW

File: client/basic-authorship/src/basic_authorship.rs [hunk 1]
```diff
@@ -52,20 +52,23 @@ use sc_proposer_metrics::MetricsLink as PrometheusMetrics;
...
+	/// phantom member to pin the `Backend`/`ProofRecording` type.
+	_phantom: PhantomData<(B, RP)>,
 	max_block_size: usize,
 }
 
-impl<A, B, C> ProposerFactory<A, B, C> {
+impl<A, B, C> ProposerFactory<A, B, C, DisableProofRecording> {
+	/// Create a new proposer factory.
+	///
+	/// Proof recording will be disabled when using proposers build by this instance to build blocks.
```
Comment (andresilva) [inline client/basic-authorship/src/basic_authorship.rs:71]:
```suggestion
	/// Proof recording will be disabled when using proposers built by this instance to build blocks.
```

File: client/basic-authorship/src/basic_authorship.rs [hunk 2]
```diff
@@ -91,7 +117,7 @@ impl<A, B, C> ProposerFactory<A, B, C> {
 	}
 }
 
-impl<B, Block, C, A> ProposerFactory<A, B, C>
+impl<B, Block, C, A, RP> ProposerFactory<A, B, C, RP>
```
Comment (andresilva) [inline client/basic-authorship/src/basic_authorship.rs:120]: nitpick I guess this is named `RP` from `RecordProof` but probably `PR` would be better (`ProofRecording`).

**Review by andresilva — APPROVED** [review]

File: client/basic-authorship/src/basic_authorship.rs (hunk 2 above)
Comment (bkchr) [reply client/basic-authorship/src/basic_authorship.rs:120]: Ha! Before I renamed the trait over the lifetime of this pr :D

**Review by cheme — COMMENTED** [review]
LGTM, a few nit and questions.

File: primitives/consensus/common/src/lib.rs [hunk 3]
```diff
@@ -89,53 +90,81 @@ pub trait Environment<B: BlockT> {
...
-	No,
+/// This is used by [`Proposer`] to signal if proof recording is enabled. This can be used by
+/// downstream users of the [`Proposer`] trait to enforce that proof recording is activated when
+/// required. The only two implementations of this trait are [`DisableProofRecording`] and
+/// [`EnableProofRecording`].
+///
+/// This trait is sealed and can not be implemented outside of this crate!
+pub trait ProofRecording: Send + Sync + private::Sealed + 'static {
+	/// The proof type that will be used internally.
+	type Proof: Send + Sync + 'static;
```
[the remark this answers is not in the document]
Comment (cheme) [reply primitives/consensus/common/src/lib.rs:118]: yeah, I just checked, I did switch to encoded_storage_proof in this branch so no need for it. Probably needed here.

File: primitives/consensus/common/src/lib.rs [hunk 4]
```diff
@@ -150,8 +179,16 @@ pub trait Proposer<B: BlockT> {
...
+	type Proposal:
+		Future<Output = Result<Proposal<B, Self::Transaction, Self::Proof>, Self::Error>>
+		+ Send
+		+ Unpin
+		+ 'static;
+	/// The supported proof recording by the implementator of this trait. See [`ProofRecording`]
+	/// for more information.
+	type ProofRecording: self::ProofRecording<Proof = Self::Proof> + Send + Sync + 'static;
+	/// The proof type used by [`Self::ProofRecording`].
+	type Proof: Send + Sync + 'static;
```
[the remark this answers is not in the document]
Comment (bkchr) [reply primitives/consensus/common/src/lib.rs:191]: Yeah I first didn't had this. I just added this to make some other code shorter ;)

File: primitives/consensus/common/src/lib.rs [hunk 5]
```diff
@@ -89,53 +90,81 @@ pub trait Environment<B: BlockT> {
...
-	/// `Yes`, record a proof.
-	Yes,
-	/// `No`, don't record any proof.
-	No,
+/// This is used by [`Proposer`] to signal if proof recording is enabled. This can be used by
+/// downstream users of the [`Proposer`] trait to enforce that proof recording is activated when
+/// required. The only two implementations of this trait are [`DisableProofRecording`] and
+/// [`EnableProofRecording`].
+///
+/// This trait is sealed and can not be implemented outside of this crate!
```
[the remark this answers is not in the document]
Comment (bkchr) [reply primitives/consensus/common/src/lib.rs:115]: For now I don't see any reason to allow this to change from the outside. Maybe in the future when there is a reason to change the proof type.

File: primitives/consensus/common/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bkchr) [reply primitives/consensus/common/src/lib.rs:118]: We are sending this stuff across futures etc. So they are needed. Maybe not `Sync`. But we can remove in the future if there is a problem.

File: bin/node/bench/src/construct.rs [hunk 6]
```diff
@@ -170,7 +170,6 @@ impl core::Benchmark for ConstructionBenchmark {
 				inherent_data_providers.create_inherent_data().expect("Create inherent data failed"),
 				Default::default(),
 				std::time::Duration::from_secs(20),
-				RecordProof::Yes,
```
[the remark this answers is not in the document]
Comment (bkchr) [reply bin/node/bench/src/construct.rs:173]: There is no reason this ever used proof recording.

File: bin/node/cli/src/service.rs [hunk 7]
```diff
@@ -611,7 +610,6 @@ mod tests {
 						inherent_data,
 						digest,
 						std::time::Duration::from_secs(1),
-						RecordProof::Yes,
```
[the remark this answers is not in the document]
Comment (bkchr) [reply bin/node/cli/src/service.rs:614]: Same

File: bin/node/bench/src/construct.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (cheme) [reply bin/node/bench/src/construct.rs:173]: I don't know, maybe this was using proof recording on purpose to have worst case scenario bench, but I really don't know.

File: primitives/consensus/common/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (cheme) [reply primitives/consensus/common/src/lib.rs:115]: I was a bit curious, but it is harmless anyway.

**Review by cheme — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Pull request merged**